File: scripts/download-ide-data.ts
/**
 * IDE Chile Data Downloader
 * Downloads all available geospatial data from Chilean government ArcGIS services
 * and prepares it for upload to Turso database
 */

import * as fs from 'fs';
import * as path from 'path';

// Service definitions
interface IDEService {
  id: string;
  name: string;
  baseUrl: string;
  layers?: { id: number; name: string; geometryType?: string }[];
}

const IDE_SERVICES: IDEService[] = [
  // ROADS
  { id: 'red-vial-chile', name: 'National Road Network', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/VIALIDAD/Red_Vial_Chile/MapServer', layers: [{ id: 0, name: 'Scale 1:5M' }, { id: 1, name: 'Scale 1:1.1M' }, { id: 2, name: 'Scale 1:500K' }, { id: 3, name: 'Scale 1:1K' }] },
  { id: 'catastro-vial', name: 'Road Cadastre', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/VIALIDAD/Catastro_Vial/MapServer' },
  { id: 'estado-red-vial', name: 'Paved Road Status', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/VIALIDAD/Estado_Red_Vial_Pavimentada/MapServer' },
  { id: 'red-vial-estructurante', name: 'Structural Road Network', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/VIALIDAD/Red_Vial_Estructurante/MapServer' },

  // BRIDGES
  { id: 'puentes', name: 'Bridges', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/VIALIDAD/Puentes/MapServer', layers: [{ id: 0, name: 'Puentes', geometryType: 'point' }] },
  { id: 'puentes-ruta5', name: 'Route 5 Bridges', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/VIALIDAD/Puentes_Ruta5/MapServer' },

  // INFRASTRUCTURE
  { id: 'infraestructura-vial', name: 'Road Infrastructure', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/VIALIDAD/Infraestructura_Vial/MapServer', layers: [{ id: 0, name: 'Weighing Stations' }, { id: 1, name: 'Toll Plazas' }, { id: 2, name: 'Ferries' }, { id: 3, name: 'Tunnels' }] },
  { id: 'pasos-fronterizos', name: 'Border Crossings', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/VIALIDAD/Pasos_Fronterizos/MapServer' },
  { id: 'zonas-descanso', name: 'Rest Areas', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/VIALIDAD/Zonas_de_Descanso/MapServer' },

  // HYDRAULIC (DOH)
  { id: 'embalses', name: 'Dams and Reservoirs', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/DOH/Embalses/MapServer', layers: [{ id: 0, name: 'Reservoir Cadastre' }] },
  { id: 'canales-cnr', name: 'Irrigation Canals', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/DOH/Canales_CNR/MapServer' },
  { id: 'apr', name: 'Rural Potable Water', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/DOH/APR/MapServer' },

  // STORM WATER
  { id: 'siall', name: 'Storm Water Infrastructure', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/DOH/SIALL/MapServer', layers: [{ id: 0, name: 'Collectors' }, { id: 2, name: 'Discharge Points' }, { id: 3, name: 'Chambers' }, { id: 5, name: 'Catch Basins' }, { id: 8, name: 'Master Plan Zones' }, { id: 9, name: 'Tributary Areas' }] },

  // SANITATION
  { id: 'ssr-ley20998', name: 'Rural Sanitation', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/DOH/SSR_Clasificados_Ley_20998/MapServer' },
  { id: 'ssr-contratos', name: 'Sanitation Contracts', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/DOH/SSR_Contratos/MapServer' },

  // WATER RESOURCES (DGA)
  { id: 'red-hidrometrica', name: 'Hydrometric Network', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/DGA/Red_Hidrometrica/MapServer', layers: [{ id: 0, name: 'Station Types' }] },
  { id: 'alertas-dga', name: 'Flood Alerts', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/DGA/ALERTAS/MapServer', layers: [{ id: 0, name: 'Hydrometeorological Network' }] },
  { id: 'estaciones-embalse', name: 'Reservoir Stations', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/DGA/ESTACION_EMBALSE/MapServer', layers: [{ id: 0, name: 'DGA Measurement Stations' }] },
  { id: 'acuiferos-protegidos', name: 'Protected Aquifers', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/DGA/Acuiferos_Protegidos/MapServer' },
  { id: 'areas-restriccion', name: 'Restriction Zones', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/DGA/Areas_de_Restriccion_y_Zonas_de_Prohibicion/MapServer' },
  { id: 'escasez-hidrica', name: 'Water Scarcity Decrees', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/DGA/Decretos_Escasez_Hidrica/MapServer' },
  { id: 'declaracion-agotamiento', name: 'Water Depletion', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/DGA/Declaracion_de_Agotamiento/MapServer' },
  { id: 'caudales-reserva', name: 'Reserved Flow Decrees', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/DGA/Decretos_Caudales_de_Reserva/MapServer' },
  { id: 'turberas-prohibicion', name: 'Peatland Prohibition', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/DGA/Area_prohibicion_para_drenajes_en_turberas/MapServer' },

  // BASE MAPS
  { id: 'limites', name: 'Administrative Boundaries', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/MAPA_BASE/LIMITES/MapServer' },
  { id: 'asentamientos', name: 'Settlements', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/MAPA_BASE/ASENTAMIENTOS/MapServer' },
  { id: 'red-hidrografica', name: 'Hydrographic Network', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/MAPA_BASE/RED_HIDROGRAFICA/MapServer' },
  { id: 'snaspe', name: 'Protected Areas', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/MAPA_BASE/SNASPE/MapServer' },

  // EMERGENCY
  { id: 'emergencias-vialidad', name: 'Road Emergencies', baseUrl: 'https://rest-sit.mop.gob.cl/arcgis/rest/services/VIALIDAD/Emergencias_Vialidad/MapServer' },
];

// Output directory
const OUTPUT_DIR = path.join(__dirname, '../data/ide-chile');

interface GeoJSONFeature {
  type: 'Feature';
  geometry: {
    type: string;
    coordinates: number[] | number[][] | number[][][];
  } | null;
  properties: Record<string, unknown>;
}

interface GeoJSONFeatureCollection {
  type: 'FeatureCollection';
  features: GeoJSONFeature[];
}

interface LayerInfo {
  id: number;
  name: string;
  type?: string;
  geometryType?: string;
}

interface ServiceInfo {
  layers?: LayerInfo[];
}

// Fetch with retry
async function fetchWithRetry(url: string, retries = 3): Promise<Response> {
  for (let i = 0; i < retries; i++) {
    try {
      const response = await fetch(url);
      if (response.ok) return response;
      console.log(`  Retry ${i + 1}/${retries} for ${url}`);
    } catch (error) {
      if (i === retries - 1) throw error;
      console.log(`  Retry ${i + 1}/${retries} after error`);
      await new Promise(r => setTimeout(r, 1000 * (i + 1)));
    }
  }
  throw new Error(`Failed after ${retries} retries`);
}

// Get service info (layers)
async function getServiceInfo(baseUrl: string): Promise<ServiceInfo> {
  try {
    const response = await fetchWithRetry(`${baseUrl}?f=json`);
    return await response.json() as ServiceInfo;
  } catch (error) {
    console.log(`  Could not get service info: ${error}`);
    return { layers: [] };
  }
}

// Query layer for all features (paginated)
async function queryLayer(baseUrl: string, layerId: number, maxRecords = 2000): Promise<GeoJSONFeatureCollection> {
  const allFeatures: GeoJSONFeature[] = [];
  let offset = 0;
  const pageSize = 1000;

  while (true) {
    const params = new URLSearchParams({
      where: '1=1',
      outFields: '*',
      returnGeometry: 'true',
      outSR: '4326',
      f: 'geojson',
      resultOffset: String(offset),
      resultRecordCount: String(pageSize),
    });

    try {
      const url = `${baseUrl}/${layerId}/query?${params}`;
      const response = await fetchWithRetry(url);
      const data = await response.json() as GeoJSONFeatureCollection & { exceededTransferLimit?: boolean };

      if (data.features && data.features.length > 0) {
        allFeatures.push(...data.features);
        console.log(`    Fetched ${allFeatures.length} features...`);

        if (data.features.length < pageSize || allFeatures.length >= maxRecords) {
          break;
        }
        offset += pageSize;
      } else {
        break;
      }
    } catch (error) {
      console.log(`    Error querying layer ${layerId}: ${error}`);
      break;
    }
  }

  return {
    type: 'FeatureCollection',
    features: allFeatures,
  };
}

// Download all data from a service
async function downloadService(service: IDEService): Promise<void> {
  console.log(`\nDownloading: ${service.name} (${service.id})`);

  const serviceDir = path.join(OUTPUT_DIR, service.id);
  if (!fs.existsSync(serviceDir)) {
    fs.mkdirSync(serviceDir, { recursive: true });
  }

  // Get service info to find layers
  const serviceInfo = await getServiceInfo(service.baseUrl);
  const layers = service.layers || serviceInfo.layers || [];

  if (layers.length === 0) {
    // Try layer 0 as default
    layers.push({ id: 0, name: 'default' });
  }

  const serviceData: {
    service: IDEService;
    downloadedAt: string;
    layers: Array<{
      id: number;
      name: string;
      featureCount: number;
      file: string;
    }>;
    totalFeatures: number;
  } = {
    service,
    downloadedAt: new Date().toISOString(),
    layers: [],
    totalFeatures: 0,
  };

  for (const layer of layers) {
    // Skip group layers (they don't have geometry)
    if (layer.type === 'Group Layer') continue;

    console.log(`  Layer ${layer.id}: ${layer.name}`);

    try {
      const geojson = await queryLayer(service.baseUrl, layer.id);

      if (geojson.features.length > 0) {
        const filename = `layer_${layer.id}.geojson`;
        const filepath = path.join(serviceDir, filename);
        fs.writeFileSync(filepath, JSON.stringify(geojson, null, 2));

        serviceData.layers.push({
          id: layer.id,
          name: layer.name,
          featureCount: geojson.features.length,
          file: filename,
        });
        serviceData.totalFeatures += geojson.features.length;

        console.log(`    Saved ${geojson.features.length} features to ${filename}`);
      } else {
        console.log(`    No features found`);
      }
    } catch (error) {
      console.log(`    Error: ${error}`);
    }

    // Small delay between layers
    await new Promise(r => setTimeout(r, 500));
  }

  // Save service metadata
  fs.writeFileSync(
    path.join(serviceDir, 'metadata.json'),
    JSON.stringify(serviceData, null, 2)
  );
}

// Main function
async function main() {
  console.log('='.repeat(60));
  console.log('IDE Chile Data Downloader');
  console.log('='.repeat(60));

  // Create output directory
  if (!fs.existsSync(OUTPUT_DIR)) {
    fs.mkdirSync(OUTPUT_DIR, { recursive: true });
  }

  const summary: Array<{
    id: string;
    name: string;
    features: number;
    status: string;
  }> = [];

  for (const service of IDE_SERVICES) {
    try {
      await downloadService(service);

      // Read metadata to get feature count
      const metadataPath = path.join(OUTPUT_DIR, service.id, 'metadata.json');
      if (fs.existsSync(metadataPath)) {
        const metadata = JSON.parse(fs.readFileSync(metadataPath, 'utf-8'));
        summary.push({
          id: service.id,
          name: service.name,
          features: metadata.totalFeatures,
          status: 'success',
        });
      }
    } catch (error) {
      console.log(`  Failed: ${error}`);
      summary.push({
        id: service.id,
        name: service.name,
        features: 0,
        status: 'failed',
      });
    }

    // Delay between services
    await new Promise(r => setTimeout(r, 1000));
  }

  // Save summary
  const summaryPath = path.join(OUTPUT_DIR, 'download-summary.json');
  fs.writeFileSync(summaryPath, JSON.stringify({
    downloadedAt: new Date().toISOString(),
    services: summary,
    totalFeatures: summary.reduce((acc, s) => acc + s.features, 0),
  }, null, 2));

  console.log('\n' + '='.repeat(60));
  console.log('Download Summary');
  console.log('='.repeat(60));
  console.table(summary);
  console.log(`\nTotal features downloaded: ${summary.reduce((acc, s) => acc + s.features, 0)}`);
  console.log(`Data saved to: ${OUTPUT_DIR}`);
}

main().catch(console.error);

File: scripts/download-ide-data.py
#!/usr/bin/env python3
"""
IDE Chile Data Downloader
Downloads geospatial data from Chilean government ArcGIS services
Converts ESRI JSON to GeoJSON format
"""

import json
import os
import time
import urllib.request
import urllib.parse
from typing import Dict, List, Any, Optional

OUTPUT_DIR = "data/ide-chile"

# Service definitions
SERVICES = [
    # DGA - Water Resources
    {"id": "red-hidrometrica", "name": "Hydrometric Network", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/DGA/Red_Hidrometrica/MapServer", "layers": [0]},
    {"id": "alertas-dga", "name": "Flood Alerts", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/DGA/ALERTAS/MapServer", "layers": [0]},
    {"id": "estaciones-embalse", "name": "Reservoir Stations", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/DGA/ESTACION_EMBALSE/MapServer", "layers": [0]},
    {"id": "acuiferos-protegidos", "name": "Protected Aquifers", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/DGA/Acuiferos_Protegidos/MapServer", "layers": [0]},
    {"id": "areas-restriccion", "name": "Restriction Zones", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/DGA/Areas_de_Restriccion_y_Zonas_de_Prohibicion/MapServer", "layers": [0]},
    {"id": "escasez-hidrica", "name": "Water Scarcity", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/DGA/Decretos_Escasez_Hidrica/MapServer", "layers": [0]},
    {"id": "declaracion-agotamiento", "name": "Water Depletion", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/DGA/Declaracion_de_Agotamiento/MapServer", "layers": [0]},
    {"id": "caudales-reserva", "name": "Reserved Flows", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/DGA/Decretos_Caudales_de_Reserva/MapServer", "layers": [0]},
    {"id": "turberas", "name": "Peatland Prohibition", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/DGA/Area_prohibicion_para_drenajes_en_turberas/MapServer", "layers": [0]},

    # DOH - Hydraulic Works
    {"id": "embalses", "name": "Dams and Reservoirs", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/DOH/Embalses/MapServer", "layers": [0]},
    {"id": "canales-cnr", "name": "Irrigation Canals", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/DOH/Canales_CNR/MapServer", "layers": [0]},
    {"id": "apr", "name": "Rural Potable Water", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/DOH/APR/MapServer", "layers": [0]},

    # SIALL - Storm Water
    {"id": "siall-colectores", "name": "Storm Collectors", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/DOH/SIALL/MapServer", "layers": [0]},
    {"id": "siall-descargas", "name": "Storm Discharge", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/DOH/SIALL/MapServer", "layers": [2]},
    {"id": "siall-camaras", "name": "Storm Chambers", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/DOH/SIALL/MapServer", "layers": [3]},
    {"id": "siall-sumideros", "name": "Catch Basins", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/DOH/SIALL/MapServer", "layers": [5]},
    {"id": "siall-areas", "name": "Tributary Areas", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/DOH/SIALL/MapServer", "layers": [9]},

    # Sanitation
    {"id": "ssr-ley20998", "name": "Rural Sanitation", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/DOH/SSR_Clasificados_Ley_20998/MapServer", "layers": [0]},

    # VIALIDAD - Roads & Infrastructure
    {"id": "puentes", "name": "Bridges", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/VIALIDAD/Puentes/MapServer", "layers": [0]},
    {"id": "infraestructura-peajes", "name": "Toll Plazas", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/VIALIDAD/Infraestructura_Vial/MapServer", "layers": [1]},
    {"id": "infraestructura-pesaje", "name": "Weighing Stations", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/VIALIDAD/Infraestructura_Vial/MapServer", "layers": [0]},
    {"id": "infraestructura-tuneles", "name": "Tunnels", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/VIALIDAD/Infraestructura_Vial/MapServer", "layers": [3]},
    {"id": "infraestructura-balsas", "name": "Ferries", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/VIALIDAD/Infraestructura_Vial/MapServer", "layers": [2]},
    {"id": "pasos-fronterizos", "name": "Border Crossings", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/VIALIDAD/Pasos_Fronterizos/MapServer", "layers": [0]},
    {"id": "zonas-descanso", "name": "Rest Areas", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/VIALIDAD/Zonas_de_Descanso/MapServer", "layers": [0]},

    # Base Maps
    {"id": "limites-regiones", "name": "Regional Boundaries", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/MAPA_BASE/LIMITES/MapServer", "layers": [0]},
    {"id": "limites-provincias", "name": "Provincial Boundaries", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/MAPA_BASE/LIMITES/MapServer", "layers": [1]},
    {"id": "limites-comunas", "name": "Communal Boundaries", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/MAPA_BASE/LIMITES/MapServer", "layers": [2]},
    {"id": "asentamientos", "name": "Settlements", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/MAPA_BASE/ASENTAMIENTOS/MapServer", "layers": [0]},
    {"id": "snaspe", "name": "Protected Areas", "url": "https://rest-sit.mop.gob.cl/arcgis/rest/services/MAPA_BASE/SNASPE/MapServer", "layers": [0]},
]


def esri_to_geojson_geometry(esri_geom: Dict, geom_type: str) -> Optional[Dict]:
    """Convert ESRI geometry to GeoJSON geometry"""
    if not esri_geom:
        return None

    if geom_type == "esriGeometryPoint":
        if "x" in esri_geom and "y" in esri_geom:
            return {
                "type": "Point",
                "coordinates": [esri_geom["x"], esri_geom["y"]]
            }
    elif geom_type == "esriGeometryPolyline":
        if "paths" in esri_geom:
            if len(esri_geom["paths"]) == 1:
                return {
                    "type": "LineString",
                    "coordinates": esri_geom["paths"][0]
                }
            else:
                return {
                    "type": "MultiLineString",
                    "coordinates": esri_geom["paths"]
                }
    elif geom_type == "esriGeometryPolygon":
        if "rings" in esri_geom:
            if len(esri_geom["rings"]) == 1:
                return {
                    "type": "Polygon",
                    "coordinates": esri_geom["rings"]
                }
            else:
                return {
                    "type": "MultiPolygon",
                    "coordinates": [[ring] for ring in esri_geom["rings"]]
                }
    elif geom_type == "esriGeometryMultipoint":
        if "points" in esri_geom:
            return {
                "type": "MultiPoint",
                "coordinates": esri_geom["points"]
            }

    return None


def esri_to_geojson(esri_data: Dict) -> Dict:
    """Convert ESRI JSON response to GeoJSON FeatureCollection"""
    features = []
    geom_type = esri_data.get("geometryType", "")

    for esri_feature in esri_data.get("features", []):
        geometry = esri_to_geojson_geometry(
            esri_feature.get("geometry"),
            geom_type
        )

        feature = {
            "type": "Feature",
            "geometry": geometry,
            "properties": esri_feature.get("attributes", {})
        }
        features.append(feature)

    return {
        "type": "FeatureCollection",
        "features": features
    }


def query_layer(base_url: str, layer_id: int, max_records: int = 10000) -> Dict:
    """Query all features from a layer with pagination"""
    all_features = []
    offset = 0
    page_size = 1000

    while True:
        params = {
            "where": "1=1",
            "outFields": "*",
            "returnGeometry": "true",
            "outSR": "4326",
            "f": "json",
            "resultOffset": str(offset),
            "resultRecordCount": str(page_size),
        }

        url = f"{base_url}/{layer_id}/query?{urllib.parse.urlencode(params)}"

        try:
            with urllib.request.urlopen(url, timeout=30) as response:
                data = json.loads(response.read().decode())

            if "error" in data:
                print(f"    API Error: {data['error'].get('message', 'Unknown error')}")
                break

            features = data.get("features", [])
            if not features:
                break

            all_features.extend(features)
            print(f"    Fetched {len(all_features)} features...")

            if len(features) < page_size or len(all_features) >= max_records:
                break

            offset += page_size
            time.sleep(0.5)  # Rate limiting

        except Exception as e:
            print(f"    Error: {e}")
            break

    # Return in ESRI format with all features
    result = {
        "features": all_features,
        "geometryType": data.get("geometryType", "") if "data" in dir() else ""
    }

    # Get geometry type from first successful response
    if all_features and "data" in dir() and "geometryType" in data:
        result["geometryType"] = data["geometryType"]

    return result


def download_service(service: Dict) -> int:
    """Download all layers from a service"""
    print(f"\nDownloading: {service['name']} ({service['id']})")

    total_features = 0

    for layer_id in service.get("layers", [0]):
        print(f"  Layer {layer_id}:")

        try:
            esri_data = query_layer(service["url"], layer_id)

            if esri_data["features"]:
                geojson = esri_to_geojson(esri_data)

                filename = f"{service['id']}_layer{layer_id}.geojson"
                filepath = os.path.join(OUTPUT_DIR, filename)

                with open(filepath, "w") as f:
                    json.dump(geojson, f)

                count = len(geojson["features"])
                total_features += count
                print(f"    Saved {count} features to {filename}")
            else:
                print(f"    No features found")

        except Exception as e:
            print(f"    Error: {e}")

        time.sleep(1)  # Rate limiting between layers

    return total_features


def main():
    print("=" * 60)
    print("IDE Chile Data Downloader")
    print("=" * 60)

    # Create output directory
    os.makedirs(OUTPUT_DIR, exist_ok=True)

    summary = []
    total_all = 0

    for service in SERVICES:
        try:
            count = download_service(service)
            summary.append({
                "id": service["id"],
                "name": service["name"],
                "features": count,
                "status": "success"
            })
            total_all += count
        except Exception as e:
            print(f"  Failed: {e}")
            summary.append({
                "id": service["id"],
                "name": service["name"],
                "features": 0,
                "status": "failed"
            })

        time.sleep(1)  # Rate limiting between services

    # Save summary
    summary_path = os.path.join(OUTPUT_DIR, "download-summary.json")
    with open(summary_path, "w") as f:
        json.dump({
            "downloadedAt": time.strftime("%Y-%m-%dT%H:%M:%SZ"),
            "services": summary,
            "totalFeatures": total_all
        }, f, indent=2)

    print("\n" + "=" * 60)
    print("Download Summary")
    print("=" * 60)
    print(f"{'Service':<30} {'Features':>10} {'Status':<10}")
    print("-" * 52)
    for s in summary:
        print(f"{s['name'][:30]:<30} {s['features']:>10} {s['status']:<10}")
    print("-" * 52)
    print(f"{'TOTAL':<30} {total_all:>10}")
    print(f"\nData saved to: {OUTPUT_DIR}/")


if __name__ == "__main__":
    main()
